src/Connection.php

```diff
@@ -1424,12 +1439,21 @@ public function commit()
 
         $connection = $this->getWrappedConnection();
 
-        if ($this->transactionNestingLevel === 1) {
-            $result = $this->doCommit($connection);
-        } elseif ($this->nestTransactionsWithSavepoints) {
-            $this->releaseSavepoint($this->_getNestedTransactionSavePointName());
+        try {
+            if ($this->transactionNestingLevel === 1) {
+                $result = $this->doCommit($connection);
+            } elseif ($this->nestTransactionsWithSavepoints) {
+                $this->releaseSavepoint($this->_getNestedTransactionSavePointName());
+            }
+        } finally {
+            $this->updateTransactionStateAfterCommit();
```

Comment:
If the above lines are successful but something goes wrong here, then `$shouldRollback` will be `true` (right?) although no rollback should happen.

That's why I was expecting some code that takes a look at the transaction nesting level before and after the commit. If the transaction level decreased all by itself.

Looking at the code though, I don't see how it could fail, so I guess this is OK.

Reply:
It's unclear to me why you are calling this function inside a `finally` block. If the commit fails, why should the transaction nesting level decrease?

Reply:
> If the commit fails, why should the transaction nesting level decrease?

We need to cleanup the transaction state. Otherwise `isTransactionActive()` will still report true to other functions (e.g. in `disconnect()`) which is not true actually. It is covered by UniqueConstraintViolationsTest.

Reply:
Why should it no longer be considered active? If `COMMIT` fails, then the transaction will be "aborted", but it won't be auto-rolled-back, right? So to me, I'd say the transaction is still going on, but it is aborted and unusable, and to regain control, you have to use `ROLLBACK TO` or `ROLLBACK`.

That's what I understand after reading the last few sentences of  https://www.postgresql.org/docs/current/tutorial-transactions.html anyway.

Reply:
You're right, good point.

I came up with the solution where we can check the exception type and if it is a known case where the transaction is auto rolled back, we won't call ROLLBACK.

[36e23d1](https://github.com/doctrine/dbal/pull/6545/commits/36e23d1f7d76081f7bb3337b7399172a419dcafe)